### PR TITLE
Add support for OCS deployment via OCS-CI for IBM Power platform - Re…

### PR DESCRIPTION
**Pull request by gitsridhar**
This is a fresh replacement for pull request 2853. I lost my development environment I used for 2853. I gathered all changes and created this pull request. 

In this commit, I have added support for deploying OCS using OCS-CI.
The supported storage cluster yaml file is included in the package.
This config yaml file is also included in the package.
A new plugin ibm_deployment.yaml is included in the package. The corresponding deployer is marked as "powervs" and type as "upi".

All comments in 2853 have been addressed.

### CODE REVIEW

File: conf/ocsci/production_powervs_upi.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+RUN:
+  bin_dir: '/usr/local/bin'
+
+DEPLOYMENT:
+  local_storage: True
+  force_download_installer: False
+  force_download_client: False
+  ocs_csv_channel: "stable-4.6"
```
Comment (ebenahar) [inline conf/ocsci/production_powervs_upi.yaml:8]: we should not put the version in the platform conf file. For picking the version when executing the test, you should append `--ocsci-conf conf/ocs_version/ocs-4.6.yaml`

File: conf/ocsci/production_powervs_upi.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  default_latest_tag: "latest-4.6"
```
Comment (ebenahar) [inline conf/ocsci/production_powervs_upi.yaml:9]: same

File: conf/ocsci/production_powervs_upi.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ENV_DATA:
+  storage_device_sets_name: "ocs-deviceset"
+  ocs_version: '4.6'
```
Comment (ebenahar) [inline conf/ocsci/production_powervs_upi.yaml:12]: same

File: conf/ocsci/production_powervs_upi.yaml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  skip_ocp_deployment: true
```
Comment (ebenahar) [inline conf/ocsci/production_powervs_upi.yaml:13]: this should not be here, but rather used as a run-ci `--ocsci-conf` param when executing

File: conf/ocsci/production_powervs_upi.yaml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  skip_ocs_deployment: false
```
Comment (ebenahar) [inline conf/ocsci/production_powervs_upi.yaml:14]: same

File: conf/ocsci/production_powervs_upi.yaml [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  platform: 'powervs'
+  disk_pattern: '/dev/vdc'
+  deployment_type: 'upi'
+  monitoring_enabled: false
+
+REPORTING:
+  email:
+    address: "ocs-ci@redhat.com"
```
Comment (ebenahar) [inline conf/ocsci/production_powervs_upi.yaml:22]: same

File: ocs_ci/deployment/deployment.py [hunk 7]
```diff
@@ -408,97 +408,143 @@ def deploy_ocs_via_operator(self):
...
-            ] = f"{device_size}Gi"
+        if self.platform == constants.IBM_POWER_PLATFORM:
+            deviceset = [None] * 3
 
-        # set storage class to OCS default on current platform
-        if self.DEFAULT_STORAGECLASS:
-            deviceset_data['dataPVCTemplate']['spec'][
-                'storageClassName'
-            ] = self.DEFAULT_STORAGECLASS
+            for i in range(3):
```
Comment (ebenahar) [inline ocs_ci/deployment/deployment.py:423]: `3` should be replaced with a conf param

File: ocs_ci/deployment/deployment.py [hunk 8]
```diff
@@ -408,97 +408,143 @@ def deploy_ocs_via_operator(self):
...
+                deviceset_data = cluster_data['spec']['storageDeviceSets'][i]
+                device_size = int(
+                    config.ENV_DATA.get('device_size', defaults.DEVICE_SIZE)
+                )
 
-        ocs_version = float(config.ENV_DATA['ocs_version'])
+                # set size of request for storage
+                if self.platform.lower() == "powervs":
```
Comment (ebenahar) [inline ocs_ci/deployment/deployment.py:430]: I wonder if we should modify the base class deployment method and not create one under the IBM deployment class

File: ocs_ci/deployment/deployment.py [hunk 9]
```diff
@@ -408,97 +408,143 @@ def deploy_ocs_via_operator(self):
...
+                        'storageClassName'
+                    ] = self.DEFAULT_STORAGECLASS_LSO
+
+                # StorageCluster tweaks for LSO
+                if config.DEPLOYMENT.get('local_storage'):
+                    cluster_data['spec']['manageNodes'] = False
+                    cluster_data['spec']['monDataDirHostPath'] = '/var/lib/rook'
+                    deviceset_data['portable'] = False
+                    deviceset_data['dataPVCTemplate']['spec']['storageClassName'] = \
+                        self.DEFAULT_STORAGECLASS_LSO
```
Comment (ebenahar) [inline ocs_ci/deployment/deployment.py:453]: deviceset_data['dataPVCTemplate']['spec']['storageClassName'] = (
    self.DEFAULT_STORAGECLASS_LSO
)

File: ocs_ci/deployment/ibm.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class IBMDeployment(Deployment):
+    """
+    Implementation of Deploy for IBM Power architecture
+    """
+
+    def destroy_ocs(self):
```
Comment (ebenahar) [inline ocs_ci/deployment/ibm.py:27]: I think this is quite the same as https://github.com/red-hat-storage/ocs-ci/blob/2c356000575f4127782ee383d2dfc0593f04500a/ocs_ci/ocs/uninstall.py#L145

File: conf/ocsci/production_powervs_upi.yaml (hunk 4 above)
Comment (clacroix12) [reply conf/ocsci/production_powervs_upi.yaml:13]: @ebenahar you should instead use `--ocs-version 4.6`, the [README](https://github.com/red-hat-storage/ocs-ci/blob/master/conf/ocs_version/README.md) explains why.

File: conf/ocsci/production_powervs_upi.yaml (hunk 1 above)
Comment (clacroix12) [reply conf/ocsci/production_powervs_upi.yaml:8]: @ebenahar you should instead use `--ocs-version 4.6`, the [README](https://github.com/red-hat-storage/ocs-ci/blob/master/conf/ocs_version/README.md) explains why.

File: ocs_ci/deployment/ibm.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+
+        # STEP 12
+        # Remove local-storage namespace
+        spdelete = 'oc delete project local-storage --wait=true --timeout=5m'
+        logger.info(spdelete)
+        out = run_cmd(spdelete)
+        logger.info(out)
+
+        # Remove CRDs
+        # crds = 'oc delete crd backingstores.noobaa.io bucketclasses.noobaa.io \
```
Comment (vasukulkarni) [inline ocs_ci/deployment/ibm.py:200]: can you remove this unused code?

**Comment by vasukulkarni** [comment]
overall LGTM and some minor comments to fix.

File: conf/ocsci/production_powervs_upi.yaml (hunk 1 above)
Comment (ebenahar) [reply conf/ocsci/production_powervs_upi.yaml:8]: right, thanks Coady

Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:8]: Ok, i see you can invoke run-ci with multiple --ocsci-conf parameters... Sure, iIwill make that change.

File: ocs_ci/deployment/ibm.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+"""
+import logging
+
+from ocs_ci.deployment.deployment import Deployment
+
+from ocs_ci.ocs import ocp
+from ocs_ci.ocs.exceptions import (
+    CommandFailed
+)
+# from ocs_ci.ocs.resources.storage_cluster import change_noobaa_endpoints_count
```
Comment (clacroix12) [inline ocs_ci/deployment/ibm.py:14]: Remove commented import

File: ocs_ci/deployment/ibm.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+        # cephfilesystems.ceph.rook.io cephnfses.ceph.rook.io \
+        # cephobjectstores.ceph.rook.io cephobjectstoreusers.ceph.rook.io \
+        # noobaas.noobaa.io ocsinitializations.ocs.openshift.io \
+        # storageclusterinitializations.ocs.openshift.io \
+        # storageclusters.ocs.openshift.io  --wait=true --timeout=5m'
+        # logger.warning(crds)
+        # logger.info(crds)
+        # Should we delete these? If so, do it before deleting openshift_storage
+        # out = run_cmd(crds)
+        # logger.info(out)
```
Comment (clacroix12) [inline ocs_ci/deployment/ibm.py:211]: Remove commented out code

File: conf/ocsci/production_powervs_upi.yaml (hunk 2 above)
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:9]: I will use --ocs-version 4.6 parameter.

File: conf/ocsci/production_powervs_upi.yaml (hunk 1 above)
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:8]: I will use --ocs-version 4.6.

File: conf/ocsci/production_powervs_upi.yaml (hunk 3 above)
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:12]: I will use --ocs-version 4.6.

File: conf/ocsci/production_powervs_upi.yaml (hunk 4 above)
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:13]: I will use --ocs-version 4.6.

File: conf/ocsci/production_powervs_upi.yaml (hunk 5 above)
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:14]: I will use --ocs-version 4.6.

File: ocs_ci/deployment/deployment.py (hunk 8 above)
Comment (gitsridhar) [reply ocs_ci/deployment/deployment.py:430]: I will revisit this in the future, I feel there are many platform-specific code in deployment.py under if conditions, they should belong to subclass. But I will look at this restructure in the future with broader discussions.

File: ocs_ci/deployment/ibm.py (hunk 10 above)
Comment (gitsridhar) [reply ocs_ci/deployment/ibm.py:27]: When I try --teardown option while running run-ci, I do not see uninstall_ocs or ocs-ci/ocs_ci/ocs/uninstall is involved. I will look at this in future. We may not use ocs teardown much, may remove OCP itself.

File: conf/ocsci/production_powervs_upi.yaml [hunk 14]
```diff
@@ -0,0 +1,30 @@
+RUN:
```
[the remark this answers is not in the document]
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:1]: The production_powervs config file has only items that are specific to it now. And with --ocs-version added to run_ci, it is even more trimmed down now.

File: conf/ocsci/production_powervs_upi.yaml [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+  bin_dir: '/usr/local/bin'
```
[the remark this answers is not in the document]
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:2]: I could not find a tar file available for oc/kubectl to be downloaded and extracted programmatically into ./bin directory for ppc64le platform. There is another pull request Manoj Kumar is pursuing with on this topic. For now, before running run-ci, the correct version of oc will be copied from the OCP bastion server to /usr/local/bin.

File: ocs_ci/deployment/deployment.py [hunk 16]
```diff
@@ -408,97 +408,143 @@ def deploy_ocs_via_operator(self):
...
-            pv_size_list = helpers.get_pv_size(storageclass=self.DEFAULT_STORAGECLASS_LSO)
-            pv_size_list.sort()
-            deviceset_data['dataPVCTemplate']['spec']['resources']['requests'][
-                'storage'
-            ] = f"{pv_size_list[0]}"
-        else:
-            deviceset_data['dataPVCTemplate']['spec']['resources']['requests'][
-                'storage'
-            ] = f"{device_size}Gi"
+        if self.platform == constants.IBM_POWER_PLATFORM:
```
[the remark this answers is not in the document]
Comment (gitsridhar) [reply ocs_ci/deployment/deployment.py:420]: Let us proceed with a separate PR for this work. As I mentioned above, I see lots of conditional code in deployment.py that needs to be restructured at class level to move them to subclasses.

File: conf/ocsci/production_powervs_upi.yaml (hunk 1 above)
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml]: Done.

File: conf/ocsci/production_powervs_upi.yaml (hunk 2 above)
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml]: Done.

File: conf/ocsci/production_powervs_upi.yaml (hunk 3 above)
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml]: Done.

File: conf/ocsci/production_powervs_upi.yaml (hunk 4 above)
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:10]: Done.

File: conf/ocsci/production_powervs_upi.yaml (hunk 5 above)
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml:11]: Done.

File: conf/ocsci/production_powervs_upi.yaml (hunk 6 above)
Comment (gitsridhar) [reply conf/ocsci/production_powervs_upi.yaml]: Done.

File: ocs_ci/deployment/deployment.py (hunk 7 above)
Comment (gitsridhar) [reply ocs_ci/deployment/deployment.py]: Done.

File: ocs_ci/deployment/deployment.py (hunk 9 above)
Comment (gitsridhar) [reply ocs_ci/deployment/deployment.py]: Done.

File: ocs_ci/deployment/ibm.py (hunk 11 above)
Comment (gitsridhar) [reply ocs_ci/deployment/ibm.py]: Done.

File: ocs_ci/deployment/ibm.py (hunk 12 above)
Comment (gitsridhar) [reply ocs_ci/deployment/ibm.py]: Done.

File: ocs_ci/deployment/ibm.py (hunk 13 above)
Comment (gitsridhar) [reply ocs_ci/deployment/ibm.py]: Done.

File: ocs_ci/deployment/ibm.py [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+        spdelete = 'oc delete project local-storage --wait=true --timeout=5m'
+        logger.info(spdelete)
+        out = run_cmd(spdelete)
+        logger.info(out)
+
+        # STEP 13
+        catsource = 'oc delete catalogsource ocs-catalogsource -n openshift-marketplace'
+        logger.info(catsource)
+        out = run_cmd(catsource)
+        logger.info(out)
```
Comment (vasukulkarni) [inline ocs_ci/deployment/ibm.py:202]: minor nit: this still is not needed based on old comment.

**Review by vasukulkarni — APPROVED** [review]

**Review by clacroix12 — APPROVED** [review]

**Pull request merged**